node/src/network/mod.rs

```diff
@@ -0,0 +1,162 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+//
+// Copyright (c) DUSK NETWORK. All rights reserved.
+
+use std::{any, default, net::IpAddr, sync::Arc};
+
+use crate::{utils::PendingQueue, BoxedFilter, Message};
+use async_trait::async_trait;
+use kadcast::{config::Config, MessageInfo, Peer};
+use tokio::sync::RwLock;
+
+mod frame;
+
+type RoutesList<const N: usize> = [Option<PendingQueue<Message>>; N];
+type FilterList<const N: usize> = [Option<BoxedFilter>; N];
+
+pub struct Listener<const N: usize> {
+    routes: Arc<RwLock<RoutesList<N>>>,
+    filters: Arc<RwLock<FilterList<N>>>,
+}
+
+impl<const N: usize> Listener<N> {
+    fn reroute(
+        &self,
+        topic: impl Into<u8>,
+        msg: Message,
+    ) -> anyhow::Result<()> {
+        let topic = topic.into() as usize;
+
+        _ = match self.routes.try_read()?.get(topic) {
+            Some(Some(r)) => r.try_send(msg),
+            _ => Ok(()),
+        };
+
+        anyhow::Ok(())
+    }
+
+    fn call_filters(
+        &self,
+        topic: impl Into<u8>,
+        msg: Message,
+    ) -> anyhow::Result<()> {
+        let topic = topic.into() as usize;
+
+        match self.filters.try_write()?.get_mut(topic) {
+            Some(Some(f)) => f.filter(&msg),
+            _ => anyhow::Ok(()),
+        }
+    }
+}
+
+impl<const N: usize> kadcast::NetworkListen for Listener<N> {
+    fn on_message(&self, message: Vec<u8>, md: MessageInfo) {
+        // TODO: Decode message
+
+        if let Err(e) = self.call_filters(0, Message::default()) {
+            /// Discarding message
+            tracing::trace!("discard message due to {:?}", e);
+            return;
+        }
+
+        if let Err(e) = self.reroute(0, Message::default()) {
+            tracing::error!("could not dispatch {:?}", e);
+        }
+    }
+}
+
+pub struct Kadcast<const N: usize> {
+    peer: Peer,
+    routes: Arc<RwLock<RoutesList<N>>>,
+    filters: Arc<RwLock<FilterList<N>>>,
+}
+
+impl<const N: usize> Kadcast<N> {
+    pub fn new(conf: Config) -> Self {
+        const INIT: Option<PendingQueue<Message>> = None;
+        let routes = Arc::new(RwLock::new([INIT; N]));
+
+        const INIT_FN: Option<BoxedFilter> = None;
+        let filters = Arc::new(RwLock::new([INIT_FN; N]));
+
+        Kadcast {
+            routes: routes.clone(),
+            filters: filters.clone(),
+            peer: Peer::new(conf, Listener { routes, filters }).unwrap(),
+        }
+    }
+}
+
+#[async_trait]
+impl<const N: usize> crate::Network for Kadcast<N> {
+    async fn broadcast(&self, msg: &Message) -> anyhow::Result<()> {
+        // TODO: broadcast
+        self.peer.broadcast(&[0u8; 8], Some(0)).await;
```

Comment:
To perform a full broadcast, the kadcast_height must be `None`
Indeed `Some(0)` will send the message only to the closest peer (if any)

Reply:
I know. The full implementation of Kadcast API wrappers should be addressed with another issue.